src/Annotation/AbstractBuilder.php

```diff
@@ -0,0 +1,353 @@
+    {
+        $this->formFactory = $formFactory;
+        return $this;
+    }
+
+    /**
+     * Set event manager instance
+     *
+     * @param  EventManagerInterface $events
+     * @return $this
+     */
+    public function setEventManager(EventManagerInterface $events)
+    {
+        $events->setIdentifiers([
+            __CLASS__,
+            get_class($this),
+        ]);
+        (new ElementAnnotationsListener())->attach($events);
+        (new FormAnnotationsListener())->attach($events);
+        $this->events = $events;
+        return $this;
+    }
+
+    /**
+     * Retrieve form factory
+     *
+     * Lazy-loads the default form factory if none is currently set.
+     *
+     * @return Factory
+     */
+    public function getFormFactory()
+    {
+        if ($this->formFactory) {
+            return $this->formFactory;
+        }
+
+        $this->formFactory = new Factory();
+        return $this->formFactory;
+    }
+
+    /**
+     * Get event manager
+     *
+     * @return EventManagerInterface
+     */
+    public function getEventManager()
+    {
+        if (null === $this->events) {
+            $this->setEventManager(new EventManager());
+        }
+        return $this->events;
+    }
+
+    /**
+     * Creates and returns a form specification for use with a factory
+     *
+     * @param  string|object $entity Either an instance or a valid class name for an entity
+     * @throws Exception\InvalidArgumentException if $entity is not an object or class name
+     * @return ArrayObject
+     */
+    public function getFormSpecification($entity)
+    {
+        if (! is_object($entity)) {
+            if ((is_string($entity) && (! class_exists($entity))) // non-existent class
+                || (! is_string($entity)) // not an object or string
+            ) {
+                throw new Exception\InvalidArgumentException(sprintf(
+                    '%s expects an object or valid class name; received "%s"',
+                    __METHOD__,
+                    var_export($entity, 1)
+                ));
+            }
+        }
+
+        $this->entity = $entity;
+        list ($formSpec, $filterSpec) = $this->getFormSpecificationInternal($entity);
+
+        if (! isset($formSpec['input_filter'])) {
+            $formSpec['input_filter'] = $filterSpec;
+        } elseif (is_array($formSpec['input_filter'])) {
+            $formSpec['input_filter'] = ArrayUtils::merge($filterSpec->getArrayCopy(), $formSpec['input_filter']);
+        }
+
+        return $formSpec;
+    }
+
+    /**
+     * Implementation of deriving a form specification from an entity
+     *
+     * Must return an array with two elements, where the first element is the form specification and the
+     * second element is the input filter specification.
+     *
+     * @param string|object $entity
+     * @return array
+     */
+    abstract public function getFormSpecificationInternal($entity): array;
```

Comment:
```suggestion
    abstract protected function getFormSpecificationInternal($entity): array;
```
No need to expose this

Reply:
Indeed, fixed.